tracing/tracing.go

```diff
@@ -0,0 +1,233 @@
+// Copyright 2021 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracing
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	config_util "github.com/prometheus/common/config"
+	"github.com/prometheus/common/version"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/resource"
+	tracesdk "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.21.0"
+	"go.opentelemetry.io/otel/trace"
+	"go.opentelemetry.io/otel/trace/noop"
+	"google.golang.org/grpc/credentials"
+
+	"github.com/prometheus/alertmanager/config"
+)
+
+const serviceName = "alertmanager"
+
+// Manager is capable of building, (re)installing and shutting down
+// the tracer provider.
+type Manager struct {
+	logger       log.Logger
+	done         chan struct{}
+	config       config.TracingConfig
+	shutdownFunc func() error
+}
+
+// NewManager creates a new tracing manager.
+func NewManager(logger log.Logger) *Manager {
+	return &Manager{
+		logger: logger,
+		done:   make(chan struct{}),
+	}
+}
+
+// Run starts the tracing manager. It registers the global text map propagator and error handler.
+// It is blocking.
+func (m *Manager) Run() {
+	otel.SetTextMapPropagator(propagation.TraceContext{})
+	otel.SetErrorHandler(otelErrHandler(func(err error) {
+		level.Error(m.logger).Log("msg", "OpenTelemetry handler returned an error", "err", err)
+	}))
+	<-m.done
+}
+
+// ApplyConfig takes care of refreshing the tracing configuration by shutting down
+// the current tracer provider (if any is registered) and installing a new one.
+func (m *Manager) ApplyConfig(cfg *config.Config) error {
```

Comment:
This is not thread-safe with `Stop`. Will it ever happen that two goroutines can call `ApplyConfig` and `Stop` concurrently?

Reply:
Apparently not - `Stop` isn't called at all (but I see your other comment suggesting I add a call)...

This was copied from `prometheus/prometheus`, where `ApplyConfig` and `Run` are called concurrently in a run.Group, and `Stop` is called as an interrupt (sequentially). I'll see about safely calling `Stop` in `main.go`...